### PR TITLE
[Index][CodeCompletion] Fix #keyPath indexing when a type is referenced

### PR DESCRIPTION
**Pull request by rockbruno**
From [SR-9020](https://bugs.swift.org/browse/SR-9020): Type references in `#keyPath` were being resolved but were not being added to the components list, which prevented the expression from being indexed when the expression included a type.

I think referencing a type in `#keyPath` is rather common as this keyword is normally created outside the scope of the watched property (like in iOS with AVFoundation types), so I created a new component type and assigned the resolved reference to it, which fixed the indexing issue. 

It also seems like code completion was crashing when accessing inner properties, so I made a change to it in order to make expressions like `#keyPath(myInt.hashValue)` complete.

### CODE REVIEW

File: test/expr/primary/keypath/keypath-objc.swift [hunk 1]
```diff
@@ -57,7 +57,7 @@ func testKeyPath(a: A, b: B) {
   let _: String = #keyPath(A.propString)
 
   // Property of String property (which looks on NSString)
-  let _: String = #keyPath(A.propString.URLsInText)
+  let _: String = #keyPath(A.propString.urlsInText)
```
Comment (akyrtzi) [inline test/expr/primary/keypath/keypath-objc.swift]: It's not clear to me, was this change necessary because of your other changes ? I don't see how.

**Comment by akyrtzi** [comment]
@rintaro please review.

File: test/expr/primary/keypath/keypath-objc.swift (hunk 1 above)
Comment (rockbruno) [reply test/expr/primary/keypath/keypath-objc.swift]: The test started to fail stating the name of the property changed, and I assumed it wasn't the case before because the expression wasn't being indexed, although it does seem weird. Was it not supposed to happen?

Comment (akyrtzi) [reply test/expr/primary/keypath/keypath-objc.swift]: 'Indexing' them should not have any effect in compiler errors, is the case that these were not even typechecked before ?

Comment (akyrtzi) [reply test/expr/primary/keypath/keypath-objc.swift]: Are you 100% sure your changes are triggering the failure ?

Comment (rockbruno) [reply test/expr/primary/keypath/keypath-objc.swift]: You can actually reproduce this in Xcode 10 with old Swift 2 properties (sorry for the iOS example, it's the only one I remember that was a property and not a method):

```
let a = #keyPath(UIColor.redColor)
print(a)
```

It will work and return `redColor` despite the property being called `red` now. You can get it to throw the error if you get the property directly, which is the case where indexing was fine:

```
extension UIColor {
    func getPath() {
        let a = #keyPath(redColor)
        print(a)
    }
}
```

I have not fully located where the `renamed` error comes from but I'm assuming it comes from this change, isn't this where this error would be thrown? https://github.com/apple/swift/blob/df64e8a5dcd96e2b402be4445e0d8e913fedf71b/lib/Sema/TypeCheckAvailability.cpp#L2437

If that's the right place, then this specific problem isn't indexing itself but just that the key paths that included types never stopped being UnresolvedProperties, so this case was never being reached by them.

Comment (akyrtzi) [reply test/expr/primary/keypath/keypath-objc.swift]: @DougGregor @rudkx please review, with these changes properties from types references in #keypaths are properly typechecked now.

File: lib/Sema/TypeCheckExprObjC.cpp [hunk 2]
```diff
@@ -215,8 +215,16 @@ Optional<Type> TypeChecker::checkObjCKeyPathExpr(DeclContext *dc,
                diag::expr_unsupported_objc_key_path_component,
                (unsigned)kind);
       continue;
-    case KeyPathExpr::Component::Kind::OptionalWrap:
     case KeyPathExpr::Component::Kind::Property:
+    case KeyPathExpr::Component::Kind::Type: {
+      //For code completion purposes, we only care about the last expression.
```
Comment (rudkx) [inline lib/Sema/TypeCheckExprObjC.cpp]: It's not clear to me how this relates to code completion.

File: lib/Sema/TypeCheckExprObjC.cpp [hunk 3]
```diff
@@ -215,8 +215,16 @@ Optional<Type> TypeChecker::checkObjCKeyPathExpr(DeclContext *dc,
...
+      if (&component == &expr->getComponents().back()) {
+        return component.getComponentType();
+      } else {
+        continue;
```
Comment (rudkx) [inline lib/Sema/TypeCheckExprObjC.cpp]: This is also changing the behavior for `Kind::Property`, is that intentional?

**Comment by rudkx** [comment]
@swift-ci Please smoke test

File: lib/Sema/TypeCheckExprObjC.cpp (hunk 2 above)
Comment (rockbruno) [reply lib/Sema/TypeCheckExprObjC.cpp]: I might have treated this in the wrong place, but the issue I was having was that this method seems to be called twice on code completion - first to resolve the properties and then again in order to return the property that's trying to be completed, as far as I could tell. I couldn't find another case where this happened, so I assumed that code completion was the only case where the resolved enum cases would trigger here. Is it the wrong place?

The main idea for this block is that if I had something like`#keyPath(MyType.myProperty.`, I had to ignore the type and return `myProperty` for code completion to work.

File: lib/Sema/TypeCheckExprObjC.cpp (hunk 3 above)
Comment (rockbruno) [reply lib/Sema/TypeCheckExprObjC.cpp]: That's because you could in theory be reaching the keypath property from other properties too, like in `#keyPath(MyType.myProperty.hashValue)`.

**Comment by CodaFi** [comment]
It's been a year.  Sorry @rockbruno, could you update this?  I'll take a look and tag @xedin as well if you do.

**Comment by rockbruno** [comment]
Sure, I'll rebase this in the next hours.

**Comment by rockbruno** [comment]
I've rebased but there are two pieces of logic that don't exist anymore, so this might not be building. I'll be able to take a look soon.

**Comment by akyrtzi** [comment]
@nathawes please take a look.

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by rockbruno** [comment]
Sorry for the delay, there were indeed new cases to be considered after rebasing.
Should be working correctly now 👍

**Comment by nathawes** [comment]
@swift-ci please test

**Comment by rockbruno** [comment]
I missed that test case, sorry for that. Fixing it

**Comment by rockbruno** [comment]
Unfortunately, these changes aren't working anymore 😢The properties aren't getting type-checked anymore post-rebase.

I think I need to do this from scratch. If it ends up being the case, I'll open a new PR for this. Sorry!

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by CodaFi** [comment]
I'm going to close this out since it seems like the approach taken here has fizzled. I appreciate your interest in fixing this bug, it sounds like you were really close.
